File: python/006-start-marker/main.py
FILE_NAME = "input.txt"
START_SIZE = 14

def process_input(input):
    queue = []
    for i, c in enumerate(input):
        for iq, cq in enumerate(queue):
            if cq == c:
                queue = queue[iq+1:]
        queue.append(c)
        if len(queue) == START_SIZE:
            return i + 1
    return -1

def read_file(file_name):
    with open(file_name, "r") as f:
        return f.readlines()[0].strip()

if __name__ == "__main__":
    input = read_file(FILE_NAME)
    print(input)
    index = process_input(input)
    print(index)

File: go/006-start-marker/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Stack struct {
	Buffer []rune
}

func (s *Stack) Add(r rune) {
	for i, t := range s.Buffer {
		if t == r {
			s.Buffer = s.Buffer[i+1:]
		}
	}
	s.Buffer = append(s.Buffer, r)
	if len(s.Buffer) > patternSize {
		s.Buffer = s.Buffer[1:]
	}
	// fmt.Println(string(s.Buffer))
}

func (s *Stack) IsStart() bool {
	return len(s.Buffer) == patternSize
}

func findStart(input string) int {
	stack := Stack{}
	for i, r := range input {
		stack.Add(r)
		if stack.IsStart() {
			return i + 1
		}
	}
	return -1
}

func readFile(fileName string) string {
	file, _ := os.Open(fileName)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Scan()
	return scanner.Text()
}

const (
	fileName    = "input.txt"
	patternSize = 14
)

func main() {
	input := readFile(fileName)
	fmt.Println(input)
	fmt.Println(findStart(input))
}
